core/internal/filetransfer/file_transfer_default_test.go

```diff
@@ -105,3 +107,82 @@ func TestDefaultFileTransfer_Upload(t *testing.T) {
 	err = ft.Upload(task)
 	assert.NoError(t, err)
 }
+
+func TestDefaultFileTransfer_UploadNotFound(t *testing.T) {
+	fnfHandler := func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusNotFound)
+	}
+	err := uploadToServerWithHandler(t, fnfHandler)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "404")
+}
+
+func TestDefaultFileTransfer_UploadConnectionClosed(t *testing.T) {
+	closeHandler := func(w http.ResponseWriter, r *http.Request) {
+		hj, ok := w.(http.Hijacker)
+		assert.True(t, ok, "webserver doesn't support hijacking")
+		conn, _, err := hj.Hijack()
+		assert.NoError(t, err, "hijacking error")
+		conn.Close()
+	}
+	err := uploadToServerWithHandler(t, closeHandler)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "giving up after 2 attempt(s)")
+}
+
+func TestDefaultFileTransfer_UploadNoResponse(t *testing.T) {
+	err := uploadToServerWithHandler(t, func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(500 * time.Millisecond)
+	})
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "context deadline exceeded")
+}
+
+func TestDefaultFileTransfer_UploadNoServer(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
+	ft := filetransfer.NewDefaultFileTransfer(observability.NewNoOpLogger(), impatientClient())
+
+	tempFile, err := os.CreateTemp("", "")
+	assert.NoError(t, err)
+	defer os.Remove(tempFile.Name())
+
+	task := &filetransfer.Task{
+		Type: filetransfer.UploadTask,
+		Path: tempFile.Name(),
+		Url:  server.URL,
+	}
+
+	// Close the server before the upload begins.
+	server.Close()
+
+	err = ft.Upload(task)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "connection refused")
+}
+
+func uploadToServerWithHandler(t *testing.T, handler func(w http.ResponseWriter, r *http.Request)) error {
+	server := httptest.NewServer(http.HandlerFunc(handler))
+	defer server.Close()
+	ft := filetransfer.NewDefaultFileTransfer(observability.NewNoOpLogger(), impatientClient())
+
+	tempFile, err := os.CreateTemp("", "")
+	assert.NoError(t, err)
+	defer os.Remove(tempFile.Name())
+
+	task := &filetransfer.Task{
+		Type: filetransfer.UploadTask,
+		Path: tempFile.Name(),
+		Url:  server.URL,
+	}
+
+	return ft.Upload(task)
+}
+
+func impatientClient() *retryablehttp.Client {
+	client := retryablehttp.NewClient()
+	client.RetryMax = 1
+	client.RetryWaitMin = 1 * time.Millisecond
+	client.RetryWaitMax = 10 * time.Millisecond
+	client.HTTPClient.Timeout = 100 * time.Millisecond
+	return client
+}
```

Comment:
Similar to the other comment about test time. Could this use a non-retrying `http.Client` instead?

Reply:
In theory these tests were partially written to distinguish behavior between retryable and non-retryable errors. I've made that clearer by adding `giving up after 2...` and checking contains/notcontains. That might only be important for downstream code, so if you want to restrict the scope of these tests we can do that.

Reply:
`file_transfer_default.go` doesn't have any code that depends on the retry behavior; it just accepts a `retryablehttp.Client` because that's what we give it in production. But it should work for anything that implements `Do()`.

Unfortunately, `retryablehttp.Client`'s `Do` signature is different from `http.Client`'s, so this wouldn't be an easy change. On second look, since you set `RetryMax = 1`, it should be fine (or maybe it should even be `RetryMax = 0`?).